contracts/access_control/lib.rs

```diff
@@ -0,0 +1,251 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::access_control::{AccessControlError, Role, CHECK_ROLE_SELECTOR, HAS_ROLE_SELECTOR};
+pub mod traits;
+use ink_lang as ink;
+
+#[ink::contract]
+mod access_control {
+
+    use ink_lang::{codegen::EmitEvent, reflect::ContractEventBase};
+    use ink_storage::{
+        traits::{PackedLayout, SpreadAllocate, SpreadLayout},
+        Mapping,
+    };
+    use scale::{Decode, Encode};
+
+    pub const HAS_ROLE_SELECTOR: [u8; 4] = [0, 0, 0, 3];
+    pub const CHECK_ROLE_SELECTOR: [u8; 4] = [0, 0, 0, 5];
+
+    #[derive(Debug, Encode, Decode, Clone, Copy, SpreadLayout, PackedLayout, PartialEq, Eq)]
+    #[cfg_attr(
+        feature = "std",
+        derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+    )]
+    pub enum Role {
+        /// Indicates a superuser.
+        Admin(AccountId),
+        /// Indicates account can terminate a contract.
+        Owner(AccountId),
+        /// Indicates account can initialize a contract from a given code hash.
+        Initializer(Hash),
+    }
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct AccessControl {
+        /// Stores a de-facto hashset of user accounts and their roles
+        pub priviledges: Mapping<(AccountId, Role), ()>,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct RoleGranted {
+        #[ink(topic)]
+        by: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+        #[ink(topic)]
+        role: Role,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct RoleRevoked {
+        #[ink(topic)]
+        by: AccountId,
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        role: Role,
+    }
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum AccessControlError {
+        MissingRole(Role),
+    }
+
+    /// Result type    
+    pub type Result<T> = core::result::Result<T, AccessControlError>;
+    /// Event type
+    pub type Event = <AccessControl as ContractEventBase>::Type;
+
+    impl AccessControl {
+        /// Creates a new contract.
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            // This call is required in order to correctly initialize the
+            // `Mapping`s of our contract.
+            ink_lang::utils::initialize_contract(|contract| Self::new_init(contract))
+        }
+
+        /// Initializes the contract.
+        ///
+        /// caller is granted admin and owner piviledges
+        fn new_init(&mut self) {
+            let caller = Self::env().caller();
+            let this = self.env().account_id();
```

Comment:
Shouldn't it be `Self::env` due to instantiation issues?

Reply:
good point, though somehow it worked :man_shrugging:

Reply:
My guess is that it doesn't really matter, they both refer to the same code. `Self::env` can be used anywhere within the contract, while `self.env` requires `self` to be an instance. So in constructor you need to use the former, while in places where `self` is a thing you can use both.

But haven't tested that, so don't quote me on that ;)